### DISCUSSION TITLE
Different elastic modulus for each element

### DISCUSSION THREAD

**Discussion by drebbel1z** [idaholab/moose#26351] [Q&A Modules: Solid mechanics]
assuming I have a 10 by 10 mesh, with different elastic modulus for each element, is there a way to set this up in moose without assigning a different element id to each element? The time spent setting up is manageable for a 10 by 10 mesh, but setting up becomes very slow when the size of the mesh increases.

**Comment by GiudGiud** [comment]
Hello

You mean a different blockID on each element?
If you can use a function for the elastic modulus, then you can specify this per-element dependency as a spatial dependency in x,y,z using a CSV file to hold the XYZ data. See piecewiseConstrantFromCSV for that function

Guillaume

**Reply by drebbel1z** [reply]
Yes. I meant blockID for each element. For context, I have a stochastic field for my elastic moduli. So ideally I would want a different element modulus for each quadrature point. Since I didn't know how to do that in moose, I have decided to do a different elastic modulus per element.

From the source code, ComputeIsotropicElasticityTensor material does not take youngs_modulus as a function but as a Real, so this will mean that I will still have to specify a material for every element. The problem is that when I do this, it takes a long time to set up for large meshes. Please let me know if I am missing something.

Looking at the ComputeIsotropicElasticityTensor code, I was thinking maybe the youngs_modulus argument can be changed to a function instead of Real. This way when computeQpElasticityTensor() is called, I can get the coordinates of the _qp and use a PiecewiseBilinear function to compute the young's modulus based on the coordinates of the _qp. After that the computed young's modulus can be used to calculate _Cijkl. 

I am hoping there is already a way to do this.

**Reply by GiudGiud** [reply]
Different block ID is one solution but it s not the only one.
If you make the elastic moduli a function of space, you can re-create this "one value per element" through the spatial dependence.

Is ComputeIsotropyElasticityTensor the only material that works here? I know there are quite a few materials with slight variations
I see this one that uses a material property?
https://mooseframework.inl.gov/source/materials/ADComputeVariableIsotropicElasticityTensor.html
A material property can be set with a Function (GenericFunctionMaterial for example) so that can be the solution?

Otherwise modifying a Real to be a Function reasonably difficult.

You could also just make a custom material much like ComputeIsotropyElasticityTensor that reads in  data tabulated by element.

**Reply by drebbel1z** [reply]
I wrote a material that takes a pieceiwisebilinear function and interpolates the elastic modulus at the qp. so all good now. Thanks for the pointers to the piecewise functions. They were helpful.
